cluster.go

```diff
@@ -0,0 +1,316 @@
+	atomic.StoreUint32(&c._reload, 1)
+}
+
+// Find the next untried address
+func (c *ClusterClient) findNextAddr(tried map[string]struct{}) string {
+	for addr := range c.addrs {
+		if _, ok := tried[addr]; !ok {
+			return addr
+		}
+	}
+	return ""
+}
+
+// Fetch slot information
+func (c *ClusterClient) fetchClusterSlots(addr string) ([]ClusterSlotInfo, error) {
+	client := NewClient(c.opt.ClientOptions(addr))
+	defer client.Close()
+
+	return client.ClusterSlots().Result()
+}
+
+// Update slot information, populate slots
+func (c *ClusterClient) update(infos []ClusterSlotInfo) {
+	for _, info := range infos {
+		for i := info.Start; i <= info.End; i++ {
+			c.slots[i] = info.Addrs
+		}
+
+		for _, addr := range info.Addrs {
+			c.addrs[addr] = struct{}{}
+		}
+	}
+}
+
+//------------------------------------------------------------------------------
+
+var errNoAddrs = errors.New("redis: no addresses")
+
+type ClusterOptions struct {
+	// A seed-list of host:port addresses of known cluster nodes
+	Addrs []string
+
+	// An optional password
+	Password string
+
+	// The maximum number of MOVED/ASK redirects to follow, before
+	// giving up. Default: 16
+	MaxRedirects int
+
+	// The maximum number of open connections to cluster nodes. Default: 10.
+	// WARNING: Each connection maintains its own connection pool. The maximum
+	// possible number of socket connections is therefore MaxConns x PoolSize
+	MaxConns int
+
+	// The maximum number of TCP sockets per connection. Default: 5
+	PoolSize int
+
+	// Timeout settings
+	DialTimeout, ReadTimeout, WriteTimeout, IdleTimeout time.Duration
+}
+
+func (opt *ClusterOptions) getPoolSize() int {
+	if opt.PoolSize < 1 {
+		return 5
+	}
+	return opt.PoolSize
+}
+
+func (opt *ClusterOptions) getMaxConns() int {
+	if opt.MaxConns < 1 {
+		return 10
+	}
+	return opt.PoolSize
+}
+
+func (opt *ClusterOptions) getDialTimeout() time.Duration {
+	if opt.DialTimeout == 0 {
+		return 5 * time.Second
+	}
+	return opt.DialTimeout
+}
+
+func (opt *ClusterOptions) getMaxRedirects() int {
+	if opt.MaxRedirects < 1 {
+		return 16
+	}
+	return opt.MaxRedirects
+}
+
+func (opt *ClusterOptions) getAddrSet() (map[string]struct{}, error) {
+	size := len(opt.Addrs)
+	if size < 1 {
+		return nil, errNoAddrs
+	}
+
+	addrs := make(map[string]struct{}, size)
+	for _, addr := range opt.Addrs {
+		addrs[addr] = struct{}{}
+	}
+	return addrs, nil
+}
+
+func (opt *ClusterOptions) ClientOptions(addr string) *Options {
+	return &Options{
+		Addr:     addr,
+		DB:       0,
+		Password: opt.Password,
+
+		DialTimeout:  opt.getDialTimeout(),
+		ReadTimeout:  opt.ReadTimeout,
+		WriteTimeout: opt.WriteTimeout,
+
+		PoolSize:    opt.getPoolSize(),
+		IdleTimeout: opt.IdleTimeout,
+	}
+}
+
+//------------------------------------------------------------------------------
+
+const hashSlots = 16384
+
+// HashSlot returns a consistent slot number between 0 and 16383
+// for any given string key
+func HashSlot(key string) int {
```

Comment:
It's really quite useful in debugging apps, I don't insist but this implementation will never change